raiden/tests/unit/transfer/mediated_transfer/test_mediatorstate_regression.py

```diff
@@ -314,3 +319,103 @@ def test_regression_mediator_send_lock_expired_with_new_block():
         'secrethash': transfer.lock.secrethash,
     })
     assert transfer.lock.secrethash not in channel1.our_state.secrethashes_to_lockedlocks
+
+
+def test_regression_mediator_task_no_routes():
+    """ The mediator must only be cleared after the waiting transfer's lock has
+    been handled.
+
+    If a node receives a transfer to mediate, but there is no route available
+    (because there is no sufficient capacity or the partner nodes are offline),
+    and a refund is not possible, the mediator task must no be cleared,
+    otherwise followup remove expired lock messages wont be processed and the
+    nodes will get out of sync.
+    """
+
+    amount = 10
+    block_number = 5
+    target = HOP2
+    expiration = 30
+    pseudo_random_generator = random.Random()
+
+    payer_channel = factories.make_channel(
+        partner_balance=amount,
+        partner_address=UNIT_TRANSFER_SENDER,
+        token_address=UNIT_TOKEN_ADDRESS,
+    )
+    payer_route = factories.route_from_channel(payer_channel)
+
+    payer_transfer = factories.make_signed_transfer_for(
+        payer_channel,
+        amount,
+        HOP1,
+        target,
+        expiration,
+        UNIT_SECRET,
+    )
+
+    available_routes = []
+    channel_map = {
+        payer_channel.identifier: payer_channel,
+    }
+
+    init_state_change = ActionInitMediator(
+        available_routes,
+        payer_route,
+        payer_transfer,
+    )
+    initial_state = None
+    init_iteration = mediator.state_transition(
+        initial_state,
+        init_state_change,
+        channel_map,
+        pseudo_random_generator,
+        block_number,
+    )
+    msg = 'The task must not be cleared, even if there is no route to forward the transfer'
+    assert init_iteration.new_state is not None, msg
+    assert must_contain_entry(init_iteration.events, SendLockedTransfer, {}) is None
+    assert must_contain_entry(init_iteration.events, SendRefundTransfer, {}) is None
+
```

Comment:
Would be a good idea to check that `waiting_transfer` was set here.

Reply:
+1